plugins/modules/ndo_node_profile.py

```diff
@@ -0,0 +1,336 @@
+    node_profile_path = None
+
+    existing_node_profiles = mso_template.template.get("fabricResourceTemplate", {}).get("template", {}).get("nodeProfiles") or []
+
+    if name or uuid:
+        match = mso_template.get_object_by_key_value_pairs(
+            object_description,
+            existing_node_profiles,
+            [KVPair("uuid", uuid) if uuid else KVPair("name", name)],
+        )
+        if match:
+            node_profile_path = "{0}/{1}".format(path, match.index)
+            mso.existing = mso.previous = mso_template.add_template_values(set_node_setting_relation(mso, copy.deepcopy(match.details)))
+    else:
+        mso.existing = [mso_template.add_template_values(set_node_setting_relation(mso, profile)) for profile in existing_node_profiles]
+
+    if state == "present":
+        if uuid and not mso.existing:
+            mso.fail_json(msg="{0} with the UUID: '{1}' not found".format(object_description, uuid))
+
+        if node_setting:
+            fabric_policy_template = MSOTemplate(mso, "fabric_policy", node_setting.get("template"), node_setting.get("template_id"))
+            fabric_policy_template.validate_template("fabricPolicy")
+            node_setting_uuid = fabric_policy_template.get_node_settings_object(uuid=None, name=node_setting.get("name"), fail_module=True).details.get("uuid")
+
+        mso_values = dict(
+            name=name,
+            description=description,
+            nodes=nodes,
+            policy=node_setting_uuid,
+        )
+
+        if match:
+            proposed_payload = copy.deepcopy(match.details)
+            append_update_ops_data(ops, proposed_payload, node_profile_path, mso_values)
+            mso.sanitize(proposed_payload, collate=True)
+        else:
+            mso.sanitize(mso_values)
+            ops.append(dict(op="add", path="{0}/-".format(path), value=mso.sent))
+
+    elif state == "absent":
+        if match:
+            ops.append(dict(op="remove", path=node_profile_path))
+
+    if not module.check_mode and ops:
+        response = mso.request(mso_template.template_path, method="PATCH", data=ops)
+        node_profiles = response.get("fabricResourceTemplate", {}).get("template", {}).get("nodeProfiles") or []
+        match = mso_template.get_object_by_key_value_pairs(
+            object_description,
+            node_profiles,
+            [KVPair("uuid", uuid) if uuid else KVPair("name", name)],
+        )
+        if match:
+            mso.existing = mso_template.add_template_values(set_node_setting_relation(mso, match.details))  # When the state is present
+        else:
+            mso.existing = {}  # When the state is absent
+    elif module.check_mode and state != "query":  # When the state is present/absent with check mode
+        mso.existing = mso_template.add_template_values(set_node_setting_relation(mso, mso.proposed)) if state == "present" else {}
+
+    mso.exit_json()
+
+
+def set_node_setting_relation(mso, mso_dict):
```

Comment:
I created a general function in my PR for this based on a comments made related to it.
I'll need to modify it if we decide to use include template and template_id for references too..
You may need to use that function once I complete it.

Reply:
Thank you. I refactored my code to use the general function instead.